src/nodes/NodeConnectionManager.ts

```diff
@@ -1173,6 +1217,99 @@ class NodeConnectionManager {
     return address;
   }
 
+  /**
+   * Will attempt to find a connection via MDNS.
+   * @param targetNodeId Id of the node we are tying to find
+   * @param externalAddress Can be optionally provided to either concatinate the 'external' scope on a NodeAddress if the port and host match,
+   * or concat the address to the returned array if not matching NodeAddress was found.
+   */
+  @ready(new nodesErrors.ErrorNodeConnectionManagerNotRunning())
+  public findNodeLocal(
+    targetNodeId: NodeId,
+    externalAddress?: NodeAddress
+  ): Array<NodeAddress> {
+    this.logger.debug(
+      `Finding local addresses for ${nodesUtils.encodeNodeId(targetNodeId)}`,
+    );
+
+    let concatExternal = true;
+
+    let addresses: Array<NodeAddress> = [];
+
+    // First check if we already have an existing MDNS Service
+    if (this.mdns[running]) {
+      const service = this.mdns.networkServices.get(mdnsUtils.toFdqn({ name: nodesUtils.encodeNodeId(targetNodeId), type: "polykey", protocol: "udp" }));
+      if (service != null) {
+        for (const host_ of service.hosts) {
+          let host: Host;
+          switch (this.quicSocket.type) {
+            case 'ipv4':
+              if (quicUtils.isIPv4(host_)) host = host_ as any
+              else if (quicUtils.isIPv4MappedIPv6(host_)) host = quicUtils.fromIPv4MappedIPv6(host_) as any;
+              else continue;
+              break;
+            case 'ipv6':
+              if (quicUtils.isIPv6(host_)) host = host_ as any
+              else continue;
+              break;
+            case 'ipv4&ipv6':
+              host = host_ as any;
+              break;
+            default:
+              continue;
+          }
+          const scopes: Array<'local' | 'external'> = ['local'];
+          if (externalAddress?.host === host && externalAddress.port === service.port as any) {
+            scopes.push('external');
+            concatExternal = false;
+          }
+          addresses.push({
+            host: host as any,
+            port: service.port as any,
+            scopes
+          });
```

Comment:
A lot of `as any` here, This is a bit of a pet peve for me so try not to do it. If you have to cast then cast it to the type you need.

Reply:
i need to make the types in mdns itself to not use Opaue types for the service detail i think

Reply:
```
if (this.mdns != null) {
      const service = this.mdns.networkServices.get(mdnsUtils.toFqdn({ name: nodesUtils.encodeNodeId(targetNodeId), type: "polykey", protocol: "udp" }));
      if (service != null) {
        for (const host_ of service.hosts) {
          let host: string;
          switch (this.quicSocket.type) {
            case 'ipv4':
              if (quicUtils.isIPv4(host_)) host = host_;
              else if (quicUtils.isIPv4MappedIPv6(host_)) host = quicUtils.fromIPv4MappedIPv6(host_);
              else continue;
              break;
            case 'ipv6':
              if (quicUtils.isIPv6(host_)) host = host_;
              else continue;
              break;
            case 'ipv4&ipv6':
              host = host_;
              break;
            default:
              continue;
          }
          const scopes: Array<'local' | 'external'> = ['local'];
          if (externalAddress?.host === host && externalAddress.port === service.port) {
            scopes.push('external');
            concatExternal = false;
          }
          addresses.push({
            host: host as Host,
            port: service.port as Port,
            scopes
          });
          this.logger.debug(
            `found address for ${nodesUtils.encodeNodeId(targetNodeId)} at ${
              host
            }:${service.port}`,
          );
        }
      }
    }
```